instrumentation/github.com/Shopify/sarama/otelsarama/producer.go

```diff
@@ -189,41 +174,67 @@ func WrapAsyncProducer(saramaConfig *sarama.Config, p sarama.AsyncProducer, opts
 
 				p.Input() <- msg
 				if saramaConfig.Producer.Return.Successes {
+					mtx.Lock()
 					producerMessageContexts[msg.Metadata] = mc
+					mtx.Unlock()
 				} else {
 					// If returning successes isn't enabled, we just finish the
 					// span right away because there's no way to know when it will
 					// be done.
-					span.End()
+					mc.span.End()
 				}
+			}
+		}
+	}()
+
+	// spawn consumer goroutine
+	go func() {
+		// Clear all spans.
+		// Sarama will consume all the successes and errors by itself while closing,
+		// so our `Successes()` and `Errors()` may get nothing and those remaining spans
+		// cannot be closed.
+		defer func() {
+			mtx.Lock()
+			for _, mc := range producerMessageContexts {
+				mc.span.End()
+			}
+			mtx.Unlock()
+		}()
+		defer close(wrapped.successes)
+		defer close(wrapped.errors)
+		for {
+			select {
 			case msg, ok := <-p.Successes():
 				if !ok {
-					// producer was closed, so exit
-					return
+					return // producer was closed, so exit
```

Comment:
I do not remember why I am not ensuring that both `p.Successes()` and `p.Errors()` are closed before doing a `return`

I need to double-check it and leave a comment if I did it intentionally